File: pa1/rosetta.ml
(*Sources
 - https://ocaml.org/learn/tutorials/set.html
 - Used to understand folding to get min: http://rosettacode.org/wiki/Greatest_element_of_a_list#OCaml
 *)

(* Wish I hadn't used so much imperative style programming in this one, but functional is hard :( *)

module StringSet = Set.Make(String);;

(* Too many references, I know, I am ashamed *)
let nodes = ref [];; (* List of strings representing nodes in graph *)
let starts = ref [];; (* Nodes with indegree zero *)
let sorted = ref [];;
let graph = Hashtbl.create 1024;; (* Adjacency list *)
let indegrees = Hashtbl.create 1024;; (* node (string) maps to color (int) *)
let finished = ref false;;

(* Prints the passed list *)
let rec print_list lst = match lst with
	  [] -> ()
	| hd :: tl -> 
	(
		print_string hd;
		print_string "\n";
		print_list tl
	)
;;

(* Returns indegree of node *)
let indegree_of node =
	(Hashtbl.find indegrees node)
;;

(* Decrements indegree of specified node *)
let dec_indegree node = 
	(Hashtbl.replace indegrees node ((Hashtbl.find indegrees node) - 1));
	()
;;

(* Increments indegree of specified node *)
let inc_indegree node = 
	Hashtbl.replace indegrees node ((Hashtbl.find indegrees node) + 1);
	()
;;

(* Returns list with specified string removed *)
let rec remove str lst = match lst with
	  [] -> []
	| hd :: tl when hd = str -> tl
	| hd :: tl -> hd :: remove str tl
;;

(* Returns minimum element in list *)
let list_min lst = match lst with
	  [] ->  invalid_arg "List is empty."
	| hd :: tl -> List.fold_left min hd tl
;;

(* Returns list of nodes with indegree zero *)
let rec get_starts acc lst = match lst with
	  [] -> acc
	| hd :: tl ->
	(
		if ((Hashtbl.find indegrees hd) = 0) then
			hd :: (get_starts acc tl)
		else 
			get_starts acc tl
	)
;;

(* Iterate over adjacent nodes, decrementing their indegree and adding to starts if indegree 0 *)
let rec check_starts adj = match adj with
	  [] -> ()
	| hd :: tl ->
	(
		dec_indegree hd;
		if (indegree_of hd) = 0 then
		(
			starts := (hd :: !starts);
			check_starts tl;
		)
		else
			check_starts tl;
	)
;;

(* Checks if there are nodes with non-zero indegree *)
let rec graph_empty lst = match lst with
	  [] -> true
	| hd :: tl -> 
		((Hashtbl.find indegrees hd) = 0) && 
		graph_empty tl

let main () =
	try
		while true do 
			let dst = (read_line ()) and src = (read_line ()) 
			in
				(* If this is the first time src has been encountered *)
 				if not (Hashtbl.mem graph src) then 
 				(
					Hashtbl.add graph src ""; (* Add it to the graph *)
					Hashtbl.add indegrees src 0; (* Keep track of its indegree *)
					nodes := src :: !nodes; (* Add it to our list of nodes *)
				);

				(* If this is the first time src has been encountered *)
				if not (Hashtbl.mem graph dst) then
				(
					Hashtbl.add graph dst ""; (* Add it to the graph *)
					Hashtbl.add indegrees dst 0; (* Keep track of its indegree *)
					nodes := dst :: !nodes; (* Add it to our list of nodes *)
				);

				(* Add edge to graph *)
				if (Hashtbl.find graph src) = "" then
					Hashtbl.replace graph src dst (* Replace the adj list if its empty *)
				else
					Hashtbl.add graph src dst;

				(* Increment dst indegree *)
				inc_indegree dst
		done
	with _ -> 
	begin
		(* Construct set of nodes with indegree zero *)

		starts := (get_starts [] !nodes);

		if !starts = [] then 
			print_string "cycle\n"
		else (

			(* Top sort with Kahn's algorithm *)
			while not !finished do
				let src = list_min !starts
				in
					starts := (remove src !starts);
					sorted := !sorted @ [src]; 

					let adj = (Hashtbl.find_all graph src)
					in 
						if adj <> [""] then
							check_starts adj
						else ();

					if !starts = [] then
						finished := true
					else ();
			done;

			if not (graph_empty !nodes) then
				print_string "cycle\n"
			else print_list !sorted;
		)
	end
;;

main();;

File: pa1/rosetta.py
'''
A.J. Varshneya
ajv4dg@virginia.edu
CS4501 - Language Design & Implementation
1/25/16
rosetta.py

 Notes
 - Python is nice.

Links that were helpful:
- Kahn's algorithm: https://en.wikipedia.org/wiki/Topological_sorting

'''
from sys import stdin
import os 

def main():
	graph = {}
	indegrees = {}

	# populate adjacency list from stdin
	while (True):
		# read from stdin
		dst = str(stdin.readline().strip())

		#dst = str(my_file.readline().strip())

		if (dst == ''):
			break
		src = str(stdin.readline().strip())
		
		#src = str(my_file.readline().strip())

		# add nodes to graph
		if src not in graph:
			graph[src] = list()
		if dst not in graph:
			graph[dst] = list()
	
		# add edge to graph
		graph[src].append(dst)

		# keep track of indegree
		if src not in indegrees:
			indegrees[src] = 0
		if dst not in indegrees:
			indegrees[dst] = 0

		indegrees[dst] += 1

	# construct set of nodes with no incoming edges
	start_nodes = set()
	for node in indegrees:
		if indegrees[node] == 0:
			start_nodes.add(node)

	# use Kahn's algorithm to top sort
	sorted_nodes = []
	while (start_nodes):
		src = min(start_nodes)
		start_nodes.remove(src)

		sorted_nodes.append(src)

		# remove edge from selected node to each of its adjacent nodes
		for dst in graph[src]:
			indegrees[dst] -= 1

			# add dst nodes to start_node candidates set when there are no more incoming edges
			if indegrees[dst] == 0:
				start_nodes.add(dst)

		del graph[src][:]

	for src in graph:
		if graph[src]:
			print 'cycle'
			exit(1)

	# print sorted list of nodes
	for node in sorted_nodes:
		print node

if __name__ == '__main__':
	main()